pulp_ansible/app/models.py

```diff
@@ -45,44 +43,67 @@ class Meta:
         unique_together = ("namespace", "name")
 
 
+class Tag(Model):
+    """A model representing a Tag."""
+
+    name = models.CharField(max_length=32, unique=True, editable=False)
+
+
 class CollectionVersion(Content):
     """
     A content type representing a Collection.
 
+    This model is primarily designed to adhere to the data format for Collection content. That spec
+    is here: https://docs.ansible.com/ansible/devel/dev_guide/collections_galaxy_meta.html
+
     Fields:
 
-        version (models.CharField): Version string.
-        metadata: Artifact metadata in JSON format.
-        contents: Contnet items (e.g. roles, modules) provided by
-            collection in JSON format.
-        quality_score: Total version quality score,
-            floating-point number in range [0, 5].
+        authors (psql_fields.ArrayField): A list of the CollectionVersion content's authors.
+        contents (psql_fields.JSONField): A JSON field with data about the contents.
+        description (models.TextField): A short summary description of the collection.
+        docs_blob (psql_fields.JSONField): A JSON field holding the various documentation blobs in
+            the collection.
+        documentation (models.URLField): The URL to any online docs.
+        homepage (models.URLField): The URL to the homepage of the collection/project.
+        issues (models.URLField): The URL to the collection issue tracker.
+        license (psql_fields.ArrayField): A list of licenses for content inside of a collection.
+        license_file (models.CharField): The path to the license file for the collection.
```

Comment:
Storing filename in database doesn't make sense. If `license_file` is provided in collection metadata, it should be loaded from the artifact and stored as a text. What's the point of storing a name of some file within an artifact?

This field is relevant for importer, it specifies a filename where to look for custom license text. For example the same works for README file. How README file is named inside an artifact doesn't matter, what matters is a file content.


CC: @awcrosby

Reply:
This model comes from the spec, so since it's in the spec it's here.  https://docs.ansible.com/ansible/devel/dev_guide/collections_galaxy_meta.html#structure

Reply:
There are several steps in the collection import process.

User writes a human readable `galaxy.yml` file, which describes in loose format collection metadata. Then mazer processes it and builds `MANIFEST.json` file, which is machine readable and has more strict format, so that it can be deserialized in unambiguous way.

`MANIFEST.json` is used to import a collection (e.g. load collection metadata, find where content is located, find where external files like README and LICENSE are stored), so that these fields can be loaded if needed.

Keynote: You're not working directly with the spec, you're not getting the exact data defined in spec.

The `license_file` attribute returned by importer doesn't have any sense because it is useful only in context of artifact being processed by importer.

Instead importer should either return license full text loaded from that files or ignore it until custom licenses are officially supported. Since this field is in spec, I believe we'll have to support it, but in a different way.

Reply:
I'm confused. do the fields in `MANIFEST.json` correspond with the fields described here:  https://docs.ansible.com/ansible/devel/dev_guide/collections_galaxy_meta.html#structure

Reply:
@bmbouter yes those fields correspond. But those fields are relevant to the importer (for the importer to find the readme and license_file files). After the import there is no requirement to store every field as they were found in `MANIFEST.json`.

For the custom license, if `license_file` is present the importer will minimally check the file exists, but since there is no direction beyond that, I think the model should ignore the `license_file` key for now.

Reply:
Not necessarily. The majority of metadata fields are stored as is. What more important is why these fields are there. In community Galaxy contents of `MANJFEST.json` are stored as is, therefore license and readme path are stored as well, we do not process that data.

But as we started modeling this data, let's focus on use cases for this data. Most of it is informational and is relevant for user. Some of it is technical, like filenames. It gives hint for a program that works with artifacts metadata where to look for specific files.

Again. This information is not relevant for anybody except program that processes the artifact.

Reply:
@awcrosby Later we may want to store text of license in order to display it to user. As it's not supported by importer now, we should either not store this field at all or add empty text field for future.

Reply:
I agree let's keep it use case focused. The main thing I'm hearing is that those fields are only used by the importer. The importer isn't the main thing I think about as the motivator of this change. I have a use case in mind that any user who see's a field written [here](https://docs.ansible.com/ansible/devel/dev_guide/collections_galaxy_meta.html#structure) would expect it in pulp_ansible too.

As a pulp_ansible user (not a galaxy developer) does this use case make sense?

To give an analogy, pulp_rpm models all attributes about RPMs in [its serializers](https://github.com/pulp/pulp_rpm/blob/master/pulp_rpm/app/serializers.py#L35-L188).

Reply:
@bmbouter Generally this information is not reserved for importer only. At the moment the importer and partially ansible-galaxy client (former mazer) are the only tools that process this data. This data is for any tool that works with an artifact. Once the artifact is loaded some of that data becomes irrelevant.

What is the use case of storing raw filenames that are never gonna be used?

If  API client wants to get README or LICENSE contents, it may request it from API server.
If this data is not stored on server, API client should work with artifact itself, it should download an artifact and read `MANIFEST.json`.
Use case when API client has to do both: download an artifact and additionally query API to get metadata that is already in the artifact is silly.

Reply:
I feel like I understand your claim, but it doesn't speak to my concern.

I'll restate my concern. pulp_ansible should store all of the spec-defined metadata fields for content types it stores. If there wasn't a spec then this would be more debatable, but in this case there is. So since the spec defines it, pulp_ansible should store it. Please help me understand the problem with this approach.

Reply:
Can you please explain the reason for this?
I'm not against storing some extra data, it's totally fine if you store filenames along with its contents.

I'm saying that storing just filename, but not storing the filename contents, which is more relevant than the filename itself is very confusing to me.

Reply:
Sure, the idea is the model field names should follow from the specification. The goal is to avoid a situation where pulp_ansible is defining yet-another-standard of what a collection means, in terms of its data definition. I'm not going to argue in favor of these being especially useful, they aren't, but that's not for pulp_ansible to decide. If it's in the spec we should include it.

Does this make any sense? What do you both think?

Reply:
When referring to spec, its the spec of the galaxy.yml file in order for importing to occur. If in the future we decide to store custom license file contents, or model the documentation text snippets, we will be "defining" another data definition of a post-imported-collection. But of course that would be adding instead of removing fields.

Agreed it is not useful (if some downloaded a collection artifact tarball, and wanted to know the path of the custom license they could look at the MANIFEST.json rather than the pulp api results from the model). But maybe having it match what the author submitted could be ok for consistency.

Reply:
I think the key point here is: 

The document which is described in the spec (e.g. `galaxy.yml`) is not even stored in the artifact itself. 
Yes, we store data from it.
Yes, it's very similar.
But this is **NOT** the `galaxy.yml` file, which is described in the spec.

So referring to that spec in order to fully comply with it is wrong. 
This spec is only relevant for user repository and the program that builds an artifact out of it. Period.

Reply:
@cutwater ok so which fields are the ones that you feel should be removed?

Also where can I learn more about the actual spec? I'm not as familiar with the standards of galaxy/ansible, maybe I'm working from the wrong spec as your saying.

Reply:
@bmbouter I don't think we have publicly available spec for `MANIFEST.json`. It should basically be a superset of `galaxy.yml`, with slight adjustments (for example license field is only a list and never a string).

Like I said, I'm ok with keeping those fields (doesn't make sense to me, but whatever, having two extra fields is not a big deal). 

What I'm trying to say it that instead of (in addition to) having those fields we should store actual file contents, that are more useful for pulp and potentially for galaxy, because users likely to be interested in file contents more than just filenames.

@awcrosby Please confirm on README file, do we store its contents separately or it's a part of documentation blob. As for license I would like importer to return file contents along with just filename, so that we store it as something useful.

Reply:
We will store the README file's rendered html in the artifact's `docs_blob` key.

The importer can return the contents of the license file - it can be named `license_file_contents` or `custom_license`, it can be added at the same level as the `metadata` key. I'm thinking not embed the key inside `metadata` since that is validation of the MANIFEST.json `collection_info` rather than the results of processing the data.

Reply:
OK the tests are now passing and all the other funcitonality is there so I want to adjust these fields and merge. Removing the `readme` field makes sense, especially since we'll be oferring the contents themselves through the `docs_blob` key.

Having `custom_license` for the license contents of the not SPDX licenses sounds good. What should we do for the `license_file` field though? Remove it?

Reply:
as far as the model, i think removing `license_file` is fine (importer will still return it but it can just be ignored when saving to database)

Reply:
OK I'm removing `license_file` and `readme`. They are silly indeed.

Reply:
Looks good to me.